### ISSUE TITLE
[Bug] WebExtensions- First page load very slow with uBlock Origin enabled

### ISSUE THREAD

**Issue by Cheap-Skate** [mozilla-mobile/fenix#8251]
## Steps to reproduce
Install uBlock Origin into Fenix. Kill Fenix. Restart Fenix. Tap on a tab to reload a page. 

### Expected behavior
First page load should take approx same time as subsequent reloads 

### Actual behavior
First page load takes ~8 seconds on my mid-ranger vs ~2 seconds for subsequent loads of the same page

I have ~150,000 uBO filters in place which is quite a few. But Fennec handles the same number of filters with ~1 second additional first page load time, Fenix is more like 6 seconds extra time.

Slow loading of filter lists in Fenix????????????

### Device information

* Android device:  Galaxy A40, Android 9.0.0
* Fenix version: Today's

**Comment by Cheap-Skate** [comment]
A profile from Fenix of the first load of arstechnica.com with uBO enabled & ~200,000 filters, total page load time  ~10 seconds of which ~6 seconds is spent in uBO stuff on the parent process.

https://perfht.ml/2SyDkbx

Page load time excludes Fenix launch time, ie Fenix is already "hot" with the tabs tray visible & I tap the arstechnica.com tab.

Fennec with a similar uBO setup has a first page load time of ~5 seconds of which (I guess) ~2 seconds is spent in uBO starting up.

**Comment by csadilek** [comment]
GV bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1616421

**Comment by csadilek** [comment]
One of the reasons of the slow first load is a localStorage access by uBlock, which happens in Fenix but not Fennec. This was fixed in https://github.com/gorhill/uBlock/commit/da0ef9454a9275aa453aa0e9c62521415a12c4d9

**Comment by csadilek** [comment]
@Cheap-Skate The uBlock fix above was shipped in uBlock 1.25.2. At least to that extent that I saw the problem I can confirm that it's fixed. However, my first page load wasn't nearly as slow as you reported. Can you verify the fix using latest Nightly and uBlock?

You can find the uBlock version under: Settings -> Add-ons Manager -> tap on uBlock Origin -> Details - > Scroll to the bottom -> Version

Note uBlock is automatically updated but it may take a day or two.

**Comment by Cheap-Skate** [comment]
@csadilek I had to uninstall and reinstall to get uBO to update to 1.25.2 (possibly another bug?)

But with 1.25.2 my cold start time is still terrible:-

check uBO 1.25.2 is installed and "on", tick most filters to get ~190k network filters + ~100 k cosmetic filters.
kill Fenix, relaunch Fenix.
Tap mozilla.org in Top Sites
It takes me 8 seconds from tapping to page loaded.  The progress bar stalls for many seconds at about 25% complete.  If I disable uBO and repeat the cold load of mozilla.org is 1-2 seconds. 

Galaxy A40, Android 9, today's Fenix

**Comment by Amejia481** [comment]
@Cheap-Skate is https://github.com/mozilla-mobile/fenix/issues/9082 the issue that you experienced when updating?

**Comment by Cheap-Skate** [comment]
@Amejia481 no, my issue is uBO failed to auto-update, it got stuck on 1.24.2 and did not auto-update to 1.25.0 and then 1.25.2.  My Fennec & Desktop Firefox updated correctly.

IIRC uBO 1.25.0 was release 21 days ago, 1.25.2 so I should have been updated.

I just manually uninstalled Fenix uBO and re-installed and got 1.25.2.

**Comment by Amejia481** [comment]
> @Amejia481 no, my issue is uBO failed to auto-update, it got stuck on 1.24.2 and did not auto-update to 1.25.0 and then 1.25.2. My Fennec & Desktop Firefox updated correctly.
> 
> IIRC uBO 1.25.0 was release 21 days ago, 1.25.2 so I should have been updated.
> 
> I just manually uninstalled Fenix uBO and re-installed and got 1.25.2.

Which version of Fenix are you using?

**Comment by Cheap-Skate** [comment]
I update Fenix Nightly nearly every day from the Play Store. Just half an hour ago I updated to the latest (screenshot below) but I guess I have had the problem for many versions of Fenix since I have updated Fenix many times without receiving a uBO update.

![Screenshot_20200311-161446_Firefox Nightly](https://user-images.githubusercontent.com/40568887/76438894-816e3280-63b3-11ea-8641-f5efb96e347b.jpg)

**Comment by Amejia481** [comment]
It's pretty strange 🤔 the only two conditions for the updater to kick up are that the device is idle and it is not running low in space. Let's keep an eye open for the next update, please file a bug if it doesn't update after 2 days of the uBlock update. Thanks

**Comment by Amejia481** [comment]
By the way, what is that yellow box in the screenshot?😄

**Comment by Cheap-Skate** [comment]
Will do.  Is there any diagnostic stuff I can see in about:addons or about:config to see if it is attempting/failing to update.

[  It's my lifesaver app Sticky Notes by Harman Bhutani. Just a set of onscreen Post-Its in different colors, write lists on them and then throw them to the side of the screen where they sit and irritate you until you finish the list. I can't recommend it highly enough. It's free as well!

https://play.google.com/store/apps/details?id=harman.notes.sticky.hrmnbhutni&hl=en_US  ]

**Comment by agi90** [comment]
@Cheap-Skate Can you post a profile with uBlock 1.5.2 that shows slow cold startup?

**Comment by Cheap-Skate** [comment]
https://perfht.ml/3cSrthC

**Comment by agi90** [comment]
mmh. I don't see much time spent in uBlock lists in the latest profile. I tested on my S10 and S7 and performance seems to be in line with Fennec (we're a little slower, but that's across the board, not just with extensions)

I think what you're seeing is uBlock sometimes not catching the lists, which causes startup to be much slower, I'm looking at that.

see video here: https://imgur.com/a/bK0zlLY

**Comment by Cheap-Skate** [comment]
I think the the update to 1.25.2 has helped somewhat with the speed. On 1.24 it was ~8-10 seconds for a cold load with many filters, now it's ~4-5 seconds. But I still see a big difference between uBO enabled/disabled on Fenix which I don't see on Fennec. 

Most of that extra delay in Fenix is when the progress bar is about 25% across. With uBO enabled, the padlock has a line through it, the progress bar moves to ~25%, it hangs for 2-3 seconds, then the padlock line disappears and the progress bar continues. I don't see that "hang" in your imgur video. I think the "hang" duration might be variable from test to test.

I also see it on my S7 Exynos with a clean load of Fenix:-
```
S7 Exynos, Fennec vs Fenix Nightly, 
both with uBO 1.25.2 & ~190k network & ~100k cosmetic filters, 
time to cold load www.google.com, eyeball timing in seconds, best of 5 loads

                  uBO
              off     on
Fennec        1.5     1.5  (can't really see a difference by eye)
Fenix         2.0     4.0
```
there's the ~2 second extra cold load time with uBO on, the progress bar "hangs" at ~25%.

I see similar on my A40 which is a little slower than my S7 so everything takes a bit longer.

**Comment by mcomella** [comment]
Hey @Cheap-Skate – are you still seeing this? If so, could you provide a little more information to help us debug?
- Do you have any relevant preferences set in Fennec or Fenix? For example, do you have sync enabled, a custom tracking protection config, etc.? Anything in `about:config` or Settings
- Would it be possible to share your custom uBO filters so we can run with them locally? We understand these may contain personal details so we could also find a way to transfer non-publicly (i.e. outside of GitHub) and we'd also understand if you are not comfortable sharing
- Would you be able to record a page load with uBO enabled for both Fennec and Fenix? Similar to what agi90 did https://github.com/mozilla-mobile/fenix/issues/8251#issuecomment-597893599 (though if you upload it as two separate videos, we can take care of splicing them)

csadilek, agi90, & amejia481 landed a few fixes but can no longer reproduce locally so we're unfortunately blocked on making additional progress on this issue without additional info.

**Comment by Cheap-Skate** [comment]
Perhaps my expectations are too high. Can I ask what I should expect....

When I kill & relaunch Fenix + uBO + 300k filters it always takes a few seconds longer to load the first page than it does with UBO disabled. On both my phones. I presume Fenix loads the 300k filters from a text file at every launch? Is a few seconds a reasonable expectation to load all those filters?

With Fennec + uBO + 300k filters it seems more complicated. I think if I kill and immediately relaunch Fennec the first page load is fast, but if it's been out of memory for a few hours the first page load is very slow. If I disable UBO the first page load is always fast. It's like the uBO filters are cached but after a few hours out of memory that cache goes stale and has to be refreshed. (maybe not, but that's what it feels like??)

@agi90 mentioned a while ago https://github.com/mozilla-mobile/fenix/issues/8251#issuecomment-597893599  "uBlock sometimes not catching the lists" what does that mean?  
So the basic question is, how much should UBO slow down Fenix's first page load? Seconds? milliseconds?

cheers

**Comment by SS1113** [comment]
I too am facing this issue. The page loads are taking a lot of time sometimes more than 5 seconds. I'm running uBlock Origin with default filter lists/Settings and uBlock filters - Annoyances, Fanboy's Annoyances. turning off uBlock Origin does reduce the load time. But subsequent page loads after a few websites are loaded are relatively faster.
![Sceenshot_20200404-163130](https://user-images.githubusercontent.com/42088907/78432184-dfb1cc80-7664-11ea-8a2d-3441f046ab7e.png)
![Screenshot_20200404-163200](https://user-images.githubusercontent.com/42088907/78429618-28b55100-7664-11ea-98a4-d66187172a19.png)


Device: Motorola One Power, Android 10.
Build: Firefox Preview 4.2.1 (Build #20941431)

**Comment by Cheap-Skate** [comment]
* Galaxy S7 Exynos
* clean install Firefox Preview Nightly from Play Store (April 12 version)
* place shortcut to mozilla.org on desktop
* kill Fenix and tap desktop shortcut
-> cold load of mozilla.org takes 3 seconds
* install uBlock Origin from add-ons menu, do not change any uBO settings, do not even look at uBO menu
-> cold load of mozilla.org takes 5 seconds
* check uBO Filter lists page, you'll see circa 93,893 network filters +30,697 cosmetic filters
* turn on  all uBO filters except "Regions, languages" and tap Update now
* 151,778 network filters + 105,107 cosmetic filters
-> cold load of mozilla.org takes 7-8 seconds

uBO settings for this test were stock apart from number of filters.
"Settings" page all default:-
Hide placeholders... Show the number... Make use of context... Disable pre-fetching... Disable hyperlink auditing... are ticked. Nothing else ticked
Filter list page:-
Auto-update..., Parse and enforce..., Ignore generic cosmetic... ticked
And then all the filters ticked as above
My filter page:-
blank
My rules page:-
as default, no-large-media: and 7x behind-the-scenes entries
Whitelist page:-
as default, 8 entries listing from about-scheme to wyciwyg-scheme

**Comment by SS1113** [comment]
I don't know why but since I have updated to Firefox Preview 4.3 the page loading seems to have improved quite a lot. Now the pages load relatively faster compared to before even after force stopping and loading the page but there's still a bit of loading time when Fenix is force stopped and the page is loaded.

**Comment by Cheap-Skate** [comment]
Just following up I think I understand what's going on and if you guys give me the OK I'd like to close this bug.

TL;DR version

For me, after fiddling with uBlock filter lists, Fenix first page load from cold is rather slow (about 7 seconds). But after using Fenix for around 10 minutes, this goes away and subsequent cold loads are much faster (about 4 seconds). Cold loads stay fast until I fiddle with the filter lists again. Wild guess - some kind of caching is going on, and it's quite lazy.

4 seconds is pretty good, basically in normal use uBO with a lot of filters adds about 1 second to cold start time on my mid-range phone.

Longer version

Galaxy A40 (midrange phone), today's Fenix, latest uBlock Origin. Cold load of mozilla.org from a shortcut on my Home screen. 120,049 network filters + 109,065 cosmetic filters.

uBlock Origin disabled, cold load repeatably takes about 3 seconds
enable uBlock Origin
before fiddling with uBlock Origin, cold load repeatably takes about 4 seconds
open uBO dashboard, tick some more filter lists. Apply changes. Untick the same filter lists. Apply changes.
Still have 120,049 network filters + 109,065 cosmetic filters, but filter lists have been modified and then modified back again.
cold load repeatably takes about 7 seconds.
use Fenix normally for 10-15 minutes
cold load repeatably takes about 4 seconds.
it stays that way indefinitely (?) unless I fiddle with the filter lists again
it's very repeatable.

**Comment by agi90** [comment]
@Cheap-Skate yep, I think you're on the right track. uBlock will cache the lists after a few minutes of uptime so that subsequent cold starts use the cached version (which is way faster). Glad it works for you too.

**Comment by csadilek** [comment]
> Just following up I think I understand what's going on and if you guys give me the OK I'd like to close this bug.

Thanks, @Cheap-Skate for giving us an update. I will close this then. Please keep reporting any issues you find :).

**Comment by mcomella** [comment]
The results of our investigation into ublock start up times are here: https://github.com/mozilla-mobile/fenix/issues/17448#issuecomment-764875262 (I always struggle to find it so I'm relinking it here)
